### ISSUE TITLE
Reduce use of PortUtil

### ISSUE THREAD

**Issue by srdo** [jamesagnew/hapi-fhir#1325]
**Describe the issue**
I would like the HAPI FHIR tests to be more stable, and to behave reasonably when run concurrently.

Reading the Javadoc of PortUtil, it is very surprising to me that using `new ServerSocket(0)` is not good enough to find a free port. The doc mentions that allocating a free port can sometimes fail, but I haven't seen or heard of this problem on other projects, and e.g. the Linux bind() docs don't mention it as a possibility (http://man7.org/linux/man-pages/man2/bind.2.html). The only case where allocating a free port should fail, is if all ephemeral ports are busy. If this happens, PortUtil won't help either.

I think it would be better to simply ensure that we can pass 0 into any server we need to start during a test, and then provide a method on that server to get the currently bound port. This method has been used on other projects (e.g. Apache Storm), and seems to have eliminated port collisions for tests there, without having the resource leak drawback (and extra code maintenance) that PortUtil has.

I've updated a few tests to demonstrate how I would prefer the port allocation to work, see https://github.com/jamesagnew/hapi-fhir/compare/master...srdo:reduce-portutil-use?expand=1. I'm happy to make this change to the remaining tests, but if the idea is rejected up front, I'd prefer not to do unnecessary work.

The only valid use of PortUtil that I can see is for tests that want to get an unused port to check client behavior when connecting to an unbound port.

**Comment by jamesagnew** [comment]
Hi @srdo - The issue isn't that allocating a port will fail, it's that under heavy multithreading it appears that two concurrent calls to `new ServerSocket(0)` happening in separate processes can sometimes return the *same* port (which means that one is going to fail eventually). At least on Linux this seems to be the case. All of the logic in PortUtil is there in order to avoid that.

I like your idea if it consistently works, although I would speculate that ultimately it will result in concurrent calls to `new ServerSocket(0)` so it would result in the same kinds of failures.

Note that this issue never happens when running the tests from an IDE since it (at least IntelliJ) doesn't run tests concurrently. It does happen often enough though on both my Ubuntu workstation as well as the Travis CI server.

**Comment by srdo** [comment]
@jamesagnew Thanks for elaborating. I couldn't get the concurrency issue to happen on Windows by running a good number of threads all opening and closing ServerSockets on port 0, but maybe it only happens on Linux or when the processes requesting ports are different. 

I'll set up a Linux box and see if I can get it to happen for me. If not I'll try updating HAPI FHIR to use the pattern I mentioned, and will try running the tests a good number of times on Linux to see if it fails.

Just to make sure, when you've seen the issue happen, the failing code has been doing something like
```
ServerSocket socket = new ServerSocket(0);
//Use socket
```
and nothing like
```
ServerSocket socket = new ServerSocket(0);
int freePort = socket.getPort();
Server server = new Server(freePort);
```
right?

**Comment by jamesagnew** [comment]
I'm not clear on the distinction there... We only ever use PortUtil to find a free port so we can open a server on it, what else could we use the port for?

**Comment by srdo** [comment]
Sorry, I should have been more clear. What I meant was, before you had PortUtil, when you were seeing these port collisions, were you opening a socket with `new ServerSocket(0)` and then using the socket directly (the first bit of code I posted), or were you opening a new socket with `new ServerSocket(0)`, getting the port, closing the socket and then opening something else using that port (the second bit of code)?

The reason I'm asking is that the second pattern is pretty common, and is also thread unsafe, as the port you get from `new ServerSocket(0)` is only reserved as long as the socket is open.

Basically I'm asking whether before PortUtil was added you were doing
```
Server server = new Server(0);
```
and that was throwing address in use exceptions, or whether you were doing 
```
ServerSocket socket = new ServerSocket(0);
int freePort = socket.getPort();
socket.close();
Server server = new Server(freePort);
```

**Comment by jamesagnew** [comment]
ah ok, we've only ever had portutil... it's gotten more fancy over time (which has dramatically cut down on the bind errors) but we've always followed the pattern like the second thing you listed above.

Worth noting though, it really looks more like:

```java
ServerSocket socket = new ServerSocket(0);
int freePort = socket.getPort();
log.info("Obtained the free port: " + freePort);
socket.close();
Server server = new Server(freePort);
```

And looking at the logs you can directly see two JVMs logging the same port at the same millisecond. That seems to be the fundamental problem.

**Comment by srdo** [comment]
Ok, then I feel more at ease. In the code you posted it is possible that thread A logs and closes the socket, and then thread B gets a socket with the same port. I think it's worth trying the "open-server-and-read-port" pattern. Will start working on it, will open a PR when it looks stable.
